File: xaiop-sdk/nodejs/src/core/states.ts
/**
 * Stream / request lifecycle states.
 *
 * idle → connecting → streaming → completing → completed
 *                                      ↘ aborted | error
 */

export type StreamStatus =
  | "idle"
  | "connecting"
  | "streaming"
  | "completing"
  | "completed"
  | "aborted"
  | "error";

export const STREAM_STATUS = Object.freeze({
  IDLE: "idle" as const,
  CONNECTING: "connecting" as const,
  STREAMING: "streaming" as const,
  COMPLETING: "completing" as const,
  COMPLETED: "completed" as const,
  ABORTED: "aborted" as const,
  ERROR: "error" as const,
});

/** Statuses where a new send / mode / url change is allowed. */
export const STREAM_IDLE_LIKE = Object.freeze([
  STREAM_STATUS.IDLE,
  STREAM_STATUS.COMPLETED,
  STREAM_STATUS.ABORTED,
  STREAM_STATUS.ERROR,
] as const);

export function isStreamBusy(status: StreamStatus): boolean {
  return (
    status === STREAM_STATUS.CONNECTING ||
    status === STREAM_STATUS.STREAMING ||
    status === STREAM_STATUS.COMPLETING
  );
}

File: xaiop-sdk/python/src/xaiop/states.py
"""Stream / request lifecycle states."""

from __future__ import annotations

StreamStatus = str

STREAM_STATUS = {
    "IDLE": "idle",
    "CONNECTING": "connecting",
    "STREAMING": "streaming",
    "COMPLETING": "completing",
    "COMPLETED": "completed",
    "ABORTED": "aborted",
    "ERROR": "error",
}

STREAM_IDLE_LIKE: tuple[str, ...] = (
    STREAM_STATUS["IDLE"],
    STREAM_STATUS["COMPLETED"],
    STREAM_STATUS["ABORTED"],
    STREAM_STATUS["ERROR"],
)


def is_stream_busy(status: str) -> bool:
    return status in (
        STREAM_STATUS["CONNECTING"],
        STREAM_STATUS["STREAMING"],
        STREAM_STATUS["COMPLETING"],
    )
